### PR TITLE
Add dependency alternatives for OpenSuse Leap 15.1

### PR DESCRIPTION
**Pull request by tobias-klein**
This pull request addresses #130. There is a pre-condition that changes in electron-installer-common (see [this pull request](https://github.com/electron-userland/electron-installer-common/pull/40)) are first merged and released on npmjs, before this change can work.

### CODE REVIEW

File: src/dependencies.js [hunk 1]
```diff
@@ -12,13 +12,13 @@ const dependencyMap = {
   gvfs: 'gvfs-client',
   kdeCliTools: ['kde-cli-tools', 'kde-cli-tools5'],
   kdeRuntime: 'kde-runtime',
-  notify: 'libnotify',
-  nss: 'nss',
+  notify: ['libnotify', 'libnotify4'],
```
Comment (fcastilloec) [inline src/dependencies.js]:
```suggestion
  notify: 'libnotify or libnotify4',
```

File: src/dependencies.js [hunk 2]
```diff
@@ -12,13 +12,13 @@ const dependencyMap = {
...
+  nss: ['nss', 'mozilla-nss'],
```
Comment (fcastilloec) [inline src/dependencies.js]:
```suggestion
  nss: 'nss or mozilla-nss',
```

File: src/dependencies.js [hunk 3]
```diff
@@ -12,13 +12,13 @@ const dependencyMap = {
...
   trashCli: 'trash-cli',
-  uuid: 'libuuid',
+  uuid: ['libuuid', 'libuuid1'],
```
Comment (fcastilloec) [inline src/dependencies.js]:
```suggestion
  uuid: 'libuuid or libuuid1',
```

File: src/dependencies.js [hunk 4]
```diff
@@ -12,13 +12,13 @@ const dependencyMap = {
...
   xdgUtils: 'xdg-utils',
   xss: 'libXScrnSaver',
-  xtst: 'libXtst'
+  xtst: ['libXtst', 'libXtst6']
```
Comment (fcastilloec) [inline src/dependencies.js]:
```suggestion
  xtst: 'libXtst or libXtst6'
```

**Comment by tobias-klein** [comment]
> As I mentioned in the `electron-installer-common` PR, this should include the `or` operator but it might break compatibility with older RPM versions. I think the only problem I can think of will be related to the CIs, so we might have to investigate if the most popular CI use a never version of RPM. I know that Travis does if you use the Ubuntu 18.04 image.
> Until we discuss the implications of this, which @malept might have some ideas/opinion about it, we can move forward with this PR.

Thanks for the feedback! I can see how this makes sense and I just made changes to see whether this works. But after changing the arrays to 'or-ed strings' the result looks rather weird. For my generated package the list of dependencies returned by `rpm -qpR` now looks like this:

(kde-cli-tools or kde-cli-tools5 or kde-runtime or trash-cli or glib2 or gvfs-client)
gtk3
libXScrnSaver
**libXtst
libXtst6**
libcurl4
libicu60_2
**libnotify
libnotify4**
libsqlite3-0
**libuuid
libuuid1
mozilla-nss
nss
or**
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(FileDigests) <= 4.6.0-1
rpmlib(PayloadFilesHavePrefix) <= 4.0-1
rpmlib(PayloadIsXz) <= 5.2-1
rpmlib(RichDependencies) <= 4.12.0-1
xdg-utils
zlib

Seems like something else is needed?

File: src/dependencies.js [hunk 5]
```diff
@@ -12,13 +12,13 @@ const dependencyMap = {
   gvfs: 'gvfs-client',
   kdeCliTools: ['kde-cli-tools', 'kde-cli-tools5'],
   kdeRuntime: 'kde-runtime',
-  notify: 'libnotify',
-  nss: 'nss',
+  notify: 'libnotify or libnotify4',
```
Comment (fcastilloec) [inline src/dependencies.js]:
```suggestion
  notify: '(libnotify or libnotify4)',
```

File: src/dependencies.js [hunk 6]
```diff
@@ -12,13 +12,13 @@ const dependencyMap = {
...
+  nss: 'nss or mozilla-nss',
```
Comment (fcastilloec) [inline src/dependencies.js]:
```suggestion
  nss: '(nss or mozilla-nss)',
```

File: src/dependencies.js [hunk 7]
```diff
@@ -12,13 +12,13 @@ const dependencyMap = {
...
   trashCli: 'trash-cli',
-  uuid: 'libuuid',
+  uuid: 'libuuid or libuuid1',
```
Comment (fcastilloec) [inline src/dependencies.js]:
```suggestion
  uuid: '(libuuid or libuuid1)',
```

File: src/dependencies.js [hunk 8]
```diff
@@ -12,13 +12,13 @@ const dependencyMap = {
...
   xdgUtils: 'xdg-utils',
   xss: 'libXScrnSaver',
-  xtst: 'libXtst'
+  xtst: 'libXtst or libXtst6'
```
Comment (fcastilloec) [inline src/dependencies.js]:
```suggestion
  xtst: '(libXtst or libXtst6)'
```

**Comment by tobias-klein** [comment]
Thanks!
If rpm 4.13 doesn't support these Boolean style dependencies, why not model this in the code to have a fallback solution? Within one list there could be a 'default' option, which would be used as fallback in case of rpm <= 4.13. And then if it's supported, the (dep1 or dep2) syntax could be generated from the list.

**Comment by fcastilloec** [comment]
That's one of the alternatives I thought of, but I'm partial on dropping support for older rpm since it's at least 3 years old. I just want to make sure that the impact is going to be minimum and other people in this project agree with it.

**Comment by tobias-klein** [comment]
> That's one of the alternatives I thought of, but I'm partial on dropping support for older rpm since it's at least 3 years old. I just want to make sure that the impact is going to be minimum and other people in this project agree with it.

For a package creation tool I actually personally think it would make sense to still support older distributions. I'm actually currently using electron-installer-redhat to generate packages for CentOS 7 (which is soon to be replaced with CentOS 8, though), and this distribution ships with rpm 4.11.3.
However, if old rpm support is dropped I could still install the new version on old distributions if really necessary.

**Comment by tobias-klein** [comment]
I've implemented a possible solution, which involves a function returning a dynamically generated map instead of using a static map for the dependencies. The fallback in case of old rpm is always the first dependency from the corresponding dependency list. Please review.

**Comment by fcastilloec** [comment]
A quick look at your code and it seems that it won't support Suse-based distros with `rpm < 4.13` because you're returning package names that aren't compatible as the default. To that end, there's no point to create the function that you've done and simply remove support for older rpm is the best alternative since it will only work on Suse if `rpm >= 4.13`.
I think that to be able to support Suse distros and older versions of rpm simultaneously, we're going to need to provide a user setting to select the type of distro.
I rather just remove support for older rpm versions, rather than create a new user setting. To lessen any issues when removing support, we can add `rpm >= 4.13` to the default dependency list.

**Comment by fcastilloec** [comment]
Some quick research for helping with what to do. It seems that Appveyor, Travis and Circle, all have images that support `rpm >= 4.13` which makes it easy to remove support for older versions

**Comment by tobias-klein** [comment]
Well, I reverted my last changes and pushed new changes as discussed earlier (added the missing parentheses for the or-ed requirements), but now it looks like some tests are failing. Will have a look later.

**Comment by fcastilloec** [comment]
@tobias-klein Can you also change this line https://github.com/electron-userland/electron-installer-redhat/blob/master/.travis.yml#L5 to: `dist: bionic` because the default Travis distro uses an older rpm version and that's why it's failing

**Comment by malept** [comment]
Ideally, we'd be able to support both RPM < 4.13 and >= 4.13. RHEL7 uses 4.11, for example. However, I believe that would require a major refactor of the package to do correctly; for non-"or" RPMs, you'd have to build one RPM per set of dependencies. So currently, one for older Fedora/RHEL, one for older OpenSuse, and then one for RPM 4.13-compatible distros. Having a flag to build one at a time doesn't work too well for the Electron Forge workflow (where you can build and publish in one command).

An important note: if we were to require RPM 4.13, it would be a breaking change, so there would be a major version bump. I also expect to have an increase in issues asking why RPMs don't work on older distros. _(I recently changed Electron Packager to a different zip extraction library that required Windows 7 users to have certain frameworks installed, and I got support issues for months afterward.)_

**Comment by tobias-klein** [comment]
What about providing an interface for users to provide their own dependencyMap, so that with those custom dependencyMaps it would still be possible to generate working RPMs for older distributions?

**Comment by malept** [comment]
I'm generally hesitant to do that because it greatly increases the public API - that means each time Electron/Chromium/Node changes any of its Linux dependencies, it will be considered a breaking change and we'll have to bump the major version. Plus, it locks us down to a specific format _(that was not designed to be public API)_, again, unless we want to bump the major version.

**Comment by fcastilloec** [comment]
I think that removing support for older rpm version while bringing support for Suse based distros is the best way to move forward, even if it's a breaking change. This PR is ready as soon as we test with travis distro bionic, so the test can pass.

**Comment by malept** [comment]
@fcastilloec sounds like you've volunteered to handle the support requests :grinning:

**Comment by fcastilloec** [comment]
@malept :laughing: sure, as long as we deploy the new version in a couple of weeks from now. Maybe we can add a `postinstall` script (or another better way) that shows a console message when installing this package mentioning that `rpm >= 4.13` is required now.

**Comment by tobias-klein** [comment]
Ok, so I updated the travis build configuration and the tests are passing. Thanks for the guidance!

@fcastilloec Are you going to handle the editorial updates regarding the breaking change of RPM?

**Comment by malept** [comment]
> Maybe we can add a `postinstall` script

I learned today that `yarn` may not print stdout of postinstall scripts. I think the best way to deal with this is to make sure the error message is as explicit as possible, and make sure that it's highlighted in the README.

**Pull request merged**